src/twisted/internet/defer.py

```diff
@@ -198,16 +223,61 @@ def setDebugging(on):
     Deferred.debug = bool(on)
 
 
-def getDebugging():
+def getDebugging() -> bool:
     """
     Determine whether L{Deferred} debugging is enabled.
     """
     return Deferred.debug
 
 
+def _cancelledToTimedOutError(value: _T, timeout: float) -> _T:
+    """
+    A default translation function that translates L{Failure}s that are
+    L{CancelledError}s to L{TimeoutError}s.
+
+    @param value: Anything
+    @param timeout: The timeout
+
+    @raise: L{TimeoutError}
+
+    @since: 16.5
+    """
+    if isinstance(value, Failure):
+        value.trap(CancelledError)
+        raise TimeoutError(timeout, "Deferred")
+    return value
+
+
+# type note: this should be Callable[[object, ...], object] but mypy doesn't allow
+#   Callable[[object], object] is next best, but disallows valid callback signatures
+DeferredCallback = Callable[..., object]
+# type note: this should be Callable[[Failure, ...], object] but mypy doesn't allow
+#   Callable[[Failure], object] is next best, but disallows valid callback signatures
+DeferredErrback = Callable[..., object]
+
+_CallbackOrderedArguments = Tuple[object, ...]
+_CallbackKeywordArguments = Mapping[str, object]
+_Callbacks = Tuple[
+    Tuple[
+        DeferredCallback,
+        _CallbackOrderedArguments,
+        _CallbackKeywordArguments,
+    ],
+    Tuple[
+        Union[DeferredErrback, DeferredCallback],
+        _CallbackOrderedArguments,
+        _CallbackKeywordArguments,
+    ],
+]
+
+_NONE_KWARGS = MappingProxyType({})  # type: _CallbackKeywordArguments
+
 # See module docstring.
 _NO_RESULT = object()
-_CONTINUE = object()
+_CONTINUE = cast(DeferredCallback, object())
```

Comment:
an enum might be better here

```
class _State(enum.Enum):
    NO_RESULT = auto()
    CONTINUE = auto()
```

then you can use `Union[DefferredCallback, Literal[_State.CONTINUE]]`

Reply:
Nice, yes, that's better.

Reply:
@graingert `Literal` is new in Python 3.8… so importing it blows up in earlier versions.

I can't think of a correct alternative pre-3.8… I should do a hack like:

```python
try:
    from typing import Literal as _Literal
except ImportError:
    Literal = Optional
else:
    Literal = _Literal
```

Which I think will make `mypy` work fine, but lie at runtime about the type.  In this case (especially since this is an internal thing), I think that in practice that turns out to be fine…?  But it's a hack so I'm wondering if you are aware of a better option.

Reply:
Maybe `if TYPE_CHECKING` instead of `try`/`except`.

Reply:
> `Literal` is new in Python 3.8… so importing it blows up in earlier versions.

`Literal` exists in [`typing-extensions`](https://pypi.org/project/typing-extensions/) and mypy supports that. It's likely we'll want to use `typing-extensions` at some point, so this might be a good moment to add it as a dependency.